sdk/src/main/scala/com/horizen/account/api/rpc/service/EthService.scala

```diff
@@ -348,32 +386,123 @@ class EthService(val scNodeViewHolderRef: ActorRef, val nvtimeout: FiniteDuratio
-          else
-            Numeric.toHexString(code)
+      getStateViewAtTag(nodeView, tag) { tagStateView =>
+        val code = tagStateView.getCode(address.toBytes)
+        if (code == null)
+          "0x"
+        else
+          Numeric.toHexString(code)
+      }
+    }
+  }
+
+  @RpcMethod("debug_traceBlockByNumber")
+  def traceBlockByNumber(blockNumber: String): DebugTraceBlockByIdView = {
+    val currentBlockNumber = Numeric.cleanHexPrefix(blockNumber)
+    val previousBlockNumber = Numeric.cleanHexPrefix((java.lang.Long.decode(currentBlockNumber) - 1).toHexString)
+
+    applyOnAccountView { nodeView =>
+      getStateViewAtTag(nodeView, previousBlockNumber) { tagStateView =>
+        {
+          val requestedBlockId = getBlockIdByTag(nodeView, currentBlockNumber)
+          val requestedBlock = nodeView.history.getBlockById(requestedBlockId).get()
+          val transactions = requestedBlock.transactions
+
+          for (mcBlockRefData <- requestedBlock.mainchainBlockReferencesData) {
+            tagStateView.applyMainchainBlockReferenceData(mcBlockRefData).get
+          }
+
+          val evmResults = transactions.map(tx => {
+            Evm.Trace(
+              tagStateView.getStateDbHandle,
+              if (tx.getFrom == null) null
+              else tx.getFrom.bytes(),
+              if (tx.getTo == null) null
+              else tx.getTo.bytes(),
+              tx.getValue,
+              tx.getData,
+              tx.getGasLimit,
+              tx.getGasPrice,
+              null
+            )
+          })
+
+          new DebugTraceBlockByIdView(evmResults.toArray)
+        }
+      }
+    }
+  }
+
+  @RpcMethod("debug_traceTransaction")
+  def traceTransaction(transactionHash: String): DebugTraceTransactionView = {
+    val requestedTransaction = getTransactionAndReceipt(transactionHash) { (tx, receipt) =>
+      new EthereumTransactionView(receipt, tx)
+    }.orNull
+
+    val currentBlockId = Numeric.cleanHexPrefix(requestedTransaction.getBlockHash)
+    val currentBlockNumber = requestedTransaction.getBlockNumber
+    val previousBlockNumber = Numeric.cleanHexPrefix((java.lang.Long.decode(currentBlockNumber) - 1).toHexString)
+
+    val evmResult: EvmResult = new EvmResult()
+
+    applyOnAccountView { nodeView =>
+      getStateViewAtTag(nodeView, previousBlockNumber) { tagStateView =>
+        {
+          val currentBlock = nodeView.history.getBlockById(currentBlockId).get()
+          val gasPool = new GasPool(BigInteger.valueOf(currentBlock.header.gasLimit))
+          val transactions = currentBlock.transactions
+
+          for (mcBlockRefData <- currentBlock.mainchainBlockReferencesData) {
+            tagStateView.applyMainchainBlockReferenceData(mcBlockRefData).get
+          }
+
+          breakable {
+            for ((tx, i) <- transactions.zipWithIndex) {
+              if (tx.id == Numeric.cleanHexPrefix(transactionHash)) {
+                val txResult = Evm
+                  .Trace(
```

Comment:
let's discuss the usage of it.
It seems that we miss intrinsic gas and other important parts for the tx we want to trace.
It means that trace results are inconsistent to the real results.
We may add trace arg to the applyMessage do then pass it further.
the results (traceLogs) can be put to the `EthereumConsensusDataReceipt` as in memory only data.

Reply:
Agree, in libevm it already is the same thing, there is just `evmApply` with trace on or off. We need a way enable/disable tracing for MessageProcessors and to allow MessageProcessors to output trace logs.

Reply:
As was decided AFK:
1) We need to use `stateView.applyTransaction(...)` interface in case of both normal apply and trace logic. That will allow us to be applicable to any kind of MessageProcessor and replay Tx/Block exactly in the same way as it was applied to the State.
2) We need to extend BlockContext with the `Tracer` class. that may haw enabled and disabled statuses.
In case of "enabled", EvmMessagerProcessor will call EVM.Apply(....tracer=true) and will store the TraceLogResults into the Tracer. That will allow us to get them here and then construct `evmResult`.